Dockerfile

```diff
@@ -1,45 +1,86 @@
+# Install rust:
 ENV PATH="$HOME/.cargo/bin:$PATH"
-# Installs the correct toolchain version from rust-toolchain.toml and then the musl target
-RUN rustup target add $(cat rust_target.txt)
-
-# Build
-COPY crates crates
-COPY ./Cargo.toml Cargo.toml
-COPY ./Cargo.lock Cargo.lock
-RUN cargo zigbuild --bin uv --target $(cat rust_target.txt) --release
-RUN cp target/$(cat rust_target.txt)/release/uv /uv
-# TODO(konsti): Optimize binary size, with a version that also works when cross compiling
-# RUN strip --strip-all /uv
+COPY rust-toolchain.toml .
+RUN <<HEREDOC
+  # Install rustup, but skip installing a default toolchain as we only want the version from `rust-toolchain.toml`:
+  curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain none
+
+  # When rustup installs the toolchain ensure it actually uses the minimal profile, avoiding excess layer weight:
+  # https://github.com/rust-lang/rustup/issues/3805#issuecomment-2094066914
+  echo 'profile = "minimal"' >> rust-toolchain.toml
+  echo 'targets = [ "aarch64-unknown-linux-musl", "x86_64-unknown-linux-musl" ]' >> rust-toolchain.toml
+  # Add the relevant musl target triples (for a building binary with static linking):
+  # Workaround until `ensure` arrives: https://github.com/rust-lang/rustup/issues/2686#issuecomment-788825744
+  rustup show
+HEREDOC
+
+# Handle individual images differences for ARM64 / AMD64:
+FROM builder-base AS builder-arm64
+ENV CARGO_BUILD_TARGET=aarch64-unknown-linux-musl
+
+FROM builder-base AS builder-amd64
+ENV CARGO_BUILD_TARGET=x86_64-unknown-linux-musl
+
+# Build app:
+FROM builder-${TARGETARCH} AS builder-app
+COPY crates/ crates/
+COPY Cargo.toml Cargo.lock .
+ARG APP_NAME=uv
+ARG CARGO_HOME=/usr/local/cargo
+ARG RUSTFLAGS="-C strip=symbols -C relocation-model=static -C target-feature=+crt-static -C opt-level=z"
+ARG TARGETPLATFORM
+RUN \
+  --mount=type=cache,target="/root/.cache/zig",id="zig-cache" \
+  # Cache mounts (dirs for crates cache + build target):
+  # https://doc.rust-lang.org/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci
+  # CAUTION: As cargo uses multiple lock files (eg: `${CARGO_HOME}/{.global-cache,.package-cache,.package-cache-mutate}`), do not mount subdirs individually.
+  --mount=type=cache,target="${CARGO_HOME}",id="cargo-cache" \
+  # This cache mount is specific enough that you may not have any concurrent builds needing to share it, communicate that expectation explicitly:
+  --mount=type=cache,target="target/",id="cargo-target-${APP_NAME}-${TARGETPLATFORM}",sharing=locked \
+  # These are redundant as they're easily reconstructed from cache above. Use TMPFS mounts to exclude from cache mounts:
+  # TMPFS mount is a better choice than `rm -rf` command (which is risky on a cache mount that is shared across concurrent builds).
+  --mount=type=tmpfs,target="${CARGO_HOME}/registry/src" \
+  --mount=type=tmpfs,target="${CARGO_HOME}/git/checkouts" \
```

Comment:
The cache mounting here is a bit excessive, I've added some context with comments. You're not really going to need to maintain much here AFAIK, so it's mostly out of sight/mind in practice.

If you're not familiar with this feature, the default `sharing` type is `shared`, which allows other concurrent builds to share the same data for each mount by matching `id`, if no `id` is configured it implicitly uses the `target` path.

We have 3 caches here, `zig`, `cargo`, and the project specific `target/` directory.
- `zig` I'm not familiar with it's cache system. I assume it's a global cache from the path and that it's fine for concurrent writers (_aka `sharing=shared`, the implicit default_).
- `cargo` mounts the entire cargo home location as it doesn't yet have an isolated cache location.
  - This does mean if any other `Dockerfile` used the `id` of `cargo-cache` they'd both be sharing the same contents, including whatever is installed in `bin/`.
  - It'd normally not be safe for concurrent writers AFAIK, but due to the lock files cargo manages here, it's a non-issue.
- The `target/` location has an `id` assigned per `TARGETPLATFORM`, and is also isolated by `APP_NAME` (`uv`) as this is not useful to share with other `Dockerfile` unrelated to `uv` project.
  - While there shouldn't be any concurrent writers, if someone was to to do several parallel builds with `RUSTFLAGS` arg being changed, the `sharing=locked` will block them as they'd all want to compile the dependencies again (_making the cache a bit less useful, but not as wasteful as individual caches via `sharing=private` which aren't deterministic for repeat builds_).
  - A previous commit did take a different approach, where the build process was part of the original builder stage and built both targets together, instead of relying on `TARGETPLATFORM`. If you don't mind always building for both targets, they could share a bit of a common `target/release/` cache contents AFAIK. The logic for the two commands below was complicated a bit more with a bash loop (_since static builds require and explicit `--target` AFAIK, even when multiple targets are configured via `rust-toolchain.toml` / `.cargo/config.toml`?_), so I opted for keeping the logic below simple to grok by using separate `TARGETPLATFORM` caches (_these can still build concurrently as I demo'd in the PR description_).

The last two `tmpfs` mounts are rather self-explanatory. There is no value in caching these to disk.

If you do want something that looks simpler, Earthly has tooling to simplify rust builds and cache management. However `Dockerfile` is more common knowledge to have and troubleshoot/maintain, adding another tool/abstraction didn't seem worthwhile to contribute.